File: bench/cpp/ops/BenchMinMaxLoc.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchMinMaxLocConfig.hpp"

#include <cvcuda/OpMinMaxLoc.hpp>
#include <cvcuda/OpReformat.hpp>

#include <nvbench/nvbench.cuh>

template<typename T>
inline nvcv::DataType GetValueDataType()
{
    if constexpr (std::is_integral_v<T> && std::is_signed_v<T>)
    {
        return nvcv::TYPE_S32;
    }
    else if constexpr (std::is_integral_v<T> && std::is_unsigned_v<T>)
    {
        return nvcv::TYPE_U32;
    }
    else if constexpr (std::is_same_v<T, float>)
    {
        return nvcv::TYPE_F32;
    }
    else if constexpr (std::is_same_v<T, double>)
    {
        return nvcv::TYPE_F64;
    }
    else
    {
        throw std::invalid_argument("Unsupported MinMaxLoc input data type");
    }
}

template<typename T>
inline size_t GetValueDataSize()
{
    if constexpr (std::is_integral_v<T>)
    {
        return sizeof(int32_t);
    }
    else
    {
        return sizeof(T);
    }
}

template<typename T>
inline void minmaxloc(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape     = benchutils::GetShape<3, int3>(state.get_string("shape"));
    const benchutils::InputKind inputKind = benchutils::GetInputKind(state.get_string("inputKind"));
    int                         maxLocs   = benchutils::GetIntParam<int>(state, "maxLocations");
    std::string                 runChoice = state.get_string("runChoice");
    std::string                 layout    = benchutils::GetStringParam(state, "layout", "NHWC");

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";

    if (layout != "NHWC" && !isPlanar && !isFakePlanar)
    {
        state.skip("MinMaxLoc benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return;
    }
    if (inputKind != benchutils::InputKind::Tensor && layout != "NHWC")
    {
        state.skip("Planar MinMaxLoc benchmark is tensor-only");
        return;
    }

    bool runMin = runChoice == "MIN" || runChoice == "MIN_MAX";
    bool runMax = runChoice == "MAX" || runChoice == "MIN_MAX";

    if (!runMin && !runMax)
    {
        throw std::invalid_argument("runChoice must be MIN, MAX, or MIN_MAX");
    }

    // clang-format off

    nvcv::DataType valueType = GetValueDataType<T>();

    nvcv::Tensor minVal({{shape.x}, "N"}, valueType);
    nvcv::Tensor minLoc({{shape.x, maxLocs}, "NM"}, nvcv::TYPE_2S32);
    nvcv::Tensor numMin({{shape.x}, "N"}, nvcv::TYPE_S32);

    nvcv::Tensor maxVal({{shape.x}, "N"}, valueType);
    nvcv::Tensor maxLoc({{shape.x, maxLocs}, "NM"}, nvcv::TYPE_2S32);
    nvcv::Tensor numMax({{shape.x}, "N"}, nvcv::TYPE_S32);

    // clang-format on

    // R/W bandwidth rationale:
    // 1 read to find min/max + 1 read to collect their locations
    // 1 or 2 writes of min/max values, locations (2S32) and quantity (S32)
    const size_t inputBytes = static_cast<size_t>(shape.x) * shape.y * shape.z * sizeof(T);
    state.add_global_memory_reads(inputBytes * (isFakePlanar ? 3 : 2));
    state.add_global_memory_writes(shape.x * (GetValueDataSize<T>() + maxLocs * sizeof(int2) + sizeof(int))
                                       * (static_cast<int>(runMin) + static_cast<int>(runMax))
                                   + (isFakePlanar ? inputBytes : 0));

    cvcuda::MinMaxLoc op;

    const nvcv::Tensor noTensor{nullptr};

    auto runOp = [&op, &runChoice, &minVal, &minLoc, &numMin, &maxVal, &maxLoc, &numMax, &noTensor](cudaStream_t s,
                                                                                                    const auto  &src)
    {
        if (runChoice == "MIN")
        {
            op(s, src, minVal, minLoc, numMin, noTensor, noTensor, noTensor);
        }
        else if (runChoice == "MAX")
        {
            op(s, src, noTensor, noTensor, noTensor, maxVal, maxLoc, numMax);
        }
        else
        {
            op(s, src, minVal, minLoc, numMin, maxVal, maxLoc, numMax);
        }
    };

    // clang-format off

    if (inputKind == benchutils::InputKind::Tensor)
    {
        nvcv::Tensor src = (isPlanar || isFakePlanar)
                             ? nvcv::Tensor({{shape.x, 1, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<T>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, 1}, "NHWC"}, benchutils::GetDataType<T>());

        benchutils::FillTensor<T>(src, benchutils::LcgValues<T>());

        if (isFakePlanar)
        {
            nvcv::Tensor interSrc({{shape.x, shape.y, shape.z, 1}, "NHWC"}, benchutils::GetDataType<T>());
            cvcuda::Reformat reformatOp;

            benchutils::warmup_and_exec(state, BENCH_MINMAXLOC_WARMUP_ITERATIONS,
                [&runOp, &reformatOp, &src, &interSrc](cudaStream_t s) {
                    reformatOp(s, src, interSrc);
                    runOp(s, interSrc);
                });
            return;
        }

        benchutils::warmup_and_exec(state, BENCH_MINMAXLOC_WARMUP_ITERATIONS,
            [&runOp, &src](cudaStream_t s) {
                runOp(s, src);
            });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        nvcv::ImageBatchVarShape src(shape.x);

        benchutils::FillImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0},
                                      benchutils::LcgValues<T>());

        benchutils::warmup_and_exec(state, BENCH_MINMAXLOC_WARMUP_ITERATIONS,
            [&runOp, &src](cudaStream_t s) {
                runOp(s, src);
            });
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(minmaxloc, NVBENCH_TYPE_AXES(BENCH_MINMAXLOC_TYPES))
BENCH_MINMAXLOC_AXES;

File: bench/python/ops/bench_minmaxloc.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA MinMaxLoc operator benchmark - Python equivalent of BenchMinMaxLoc.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cupy as cp  # noqa: E402
import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_dtype_size,
    create_tensor,
    create_image_batch_varshape,
    create_stream_cache,
    run_benchmark,
)


def minmaxloc(state):
    """MinMaxLoc operator benchmark matching C++ BenchMinMaxLoc.cpp"""

    shape = parse_shape(state.get_string("shape"))
    dtype = get_dtype(state.get_string("InOutDataType"))
    input_kind = get_input_kind(state.get_string("inputKind"))
    max_locs = state.get_int64("maxLocations")
    run_choice = state.get_string("runChoice")
    try:
        layout = state.get_string("layout")
    except (KeyError, RuntimeError):
        layout = "NHWC"
    device_id = state.get_device()

    N, H, W = shape
    dtype_size = get_dtype_size(dtype)

    if run_choice not in ("MIN", "MAX", "MIN_MAX"):
        raise ValueError("runChoice must be MIN, MAX, or MIN_MAX")
    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip(
            "MinMaxLoc benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts"
        )
        return None
    if input_kind != "Tensor" and layout != "NHWC":
        state.skip("Planar MinMaxLoc benchmark is tensor-only")
        return None

    run_min = run_choice in ("MIN", "MIN_MAX")
    run_max = run_choice in ("MAX", "MIN_MAX")
    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"

    input_bytes = N * H * W * dtype_size
    state.add_global_memory_reads((3 if is_fake_planar else 2) * input_bytes)
    state.add_global_memory_writes(
        (int(run_min) + int(run_max)) * N * (4 + max_locs * 8 + 4)
        + (input_bytes if is_fake_planar else 0)
    )

    get_stream = create_stream_cache()

    if dtype in (cvcuda.Type.S8, cvcuda.Type.S16, cvcuda.Type.S32):
        val_dtype = cvcuda.Type.S32
    elif dtype in (cvcuda.Type.U8, cvcuda.Type.U16, cvcuda.Type.U32):
        val_dtype = cvcuda.Type.U32
    else:
        val_dtype = dtype

    val_dtype_map = {
        cvcuda.Type.U32: cp.uint32,
        cvcuda.Type.S32: cp.int32,
        cvcuda.Type.F32: cp.float32,
        cvcuda.Type.F64: cp.float64,
    }

    with cp.cuda.Device(device_id):
        min_val_data = cp.zeros((N, 1), dtype=val_dtype_map[val_dtype])
        min_val = cvcuda.as_tensor(min_val_data, "NC")

        min_loc_data = cp.zeros((N, max_locs, 2), dtype=cp.int32)
        min_loc = cvcuda.as_tensor(min_loc_data, "NMC")

        num_min_data = cp.zeros((N, 1), dtype=cp.int32)
        num_min = cvcuda.as_tensor(num_min_data, "NC")

        max_val_data = cp.zeros((N, 1), dtype=val_dtype_map[val_dtype])
        max_val = cvcuda.as_tensor(max_val_data, "NC")

        max_loc_data = cp.zeros((N, max_locs, 2), dtype=cp.int32)
        max_loc = cvcuda.as_tensor(max_loc_data, "NMC")

        num_max_data = cp.zeros((N, 1), dtype=cp.int32)
        num_max = cvcuda.as_tensor(num_max_data, "NC")

    def run_op(stream, src):
        if run_choice == "MIN":
            cvcuda.min_loc_into(min_val, min_loc, num_min, src, stream=stream)
        elif run_choice == "MAX":
            cvcuda.max_loc_into(max_val, max_loc, num_max, src, stream=stream)
        else:
            cvcuda.min_max_loc_into(
                min_val,
                min_loc,
                num_min,
                max_val,
                max_loc,
                num_max,
                src,
                stream=stream,
            )

    if input_kind == "Tensor":  # Tensor mode
        tensor_shape = (N, 1, H, W) if (is_planar or is_fake_planar) else (N, H, W, 1)
        src = create_tensor(
            tensor_shape,
            dtype,
            device_id,
            layout="NCHW" if (is_planar or is_fake_planar) else "NHWC",
            fill_mode="lcg",
        )

        if is_fake_planar:
            inter_src = create_tensor(
                (N, H, W, 1), dtype, device_id, layout="NHWC", fill_mode=0
            )

            def run_fake(launch):
                stream = get_stream(launch)
                cvcuda.reformat_into(inter_src, src, stream=stream)
                run_op(stream, inter_src)

            return run_fake

        def run(launch):
            run_op(get_stream(launch), src)

        return run

    else:  # ImageBatchVarShape mode
        if dtype == cvcuda.Type.U8:
            img_format = cvcuda.Format.U8
        elif dtype == cvcuda.Type.U16:
            img_format = cvcuda.Format.U16
        elif dtype == cvcuda.Type.U32:
            img_format = cvcuda.Format.U32
        elif dtype == cvcuda.Type.S16:
            img_format = cvcuda.Format.S16
        elif dtype == cvcuda.Type.S32:
            img_format = cvcuda.Format.S32
        else:
            img_format = cvcuda.Format.F32

        src = create_image_batch_varshape(
            (N, H, W), 0, img_format, dtype, device_id, fill_mode="lcg"
        )

        def run(launch):
            run_op(get_stream(launch), src)

        return run


if __name__ == "__main__":
    run_benchmark("minmaxloc", minmaxloc)
